dvc/remote/base.py

```diff
@@ -195,8 +201,8 @@ def get_dir_checksum(self, path_info):
         if self.cache.changed_cache_file(checksum):
             self.cache.move(from_info, to_info)
 
-        self.state.save(path_info, checksum)
-        self.state.save(to_info, checksum)
+        self.state.save(path_info, checksum, self.get_prefer_checksum_type())
```

Comment:
Does state really need to know checksum_type explicitly? Maybe we could just pass it a dict that looks something like `{"md5": "mymd5"}`? Relevant to the comment above.

Reply:
Why do you think passing a dict contains checksum and checksum_type make state doesn't know about checksum_type...

Reply:
And state needs to keep the old data about checksums and checksum types, so yes, it needs to know

Reply:
```
self.state.save(path_info, checksum, self.get_prefer_checksum_type())
compare to
self.state.save(path_info, {self.get_prefer_checksum_type(): checksum})
```

Which one do you like? Why?

Reply:
Keep packing and unpacking checksum and checksum type as a dict is not efficient and is unnecessary.
And doing this means many refactoring with the legacy code.

Reply:
But you are packing it into dict in state.save() anyway 🙂 And then you are manually updating particular keys. So why not just `.update()` already existing record with this new one? 🙂

Reply:
Packing it into dict later is better, because it will save you an unpacking when you need to in the future. 🙂 
Or in your design, you will pass multiple checksum types and checksums here?